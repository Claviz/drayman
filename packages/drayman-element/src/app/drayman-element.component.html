<ng-container *ngIf="viewTree">
    <!-- <ng-container *ngFor="let element of viewTree; trackBy: trackByFn"> -->
    <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: viewTree}"></ng-container>
    <!-- </ng-container> -->
</ng-container>
<ng-template #childrenTemplate let-children="children">
    <ng-container *ngFor="let element of children; trackBy: trackByFn">
        <ng-container *ngIf="!isHtmlElement(element.type)">
            <ax-lazy-element [ngStyle]="element.style"
                *axLazyElementDynamic="element.type; url: config?.elementUrl + element.type;"
                [options]="element.options">
            </ax-lazy-element>
        </ng-container>
        <ng-container *ngIf="isHtmlElement(element.type)">
            <ng-container [ngSwitch]="element.type">
                <!--ELEMENTS-START-->
                <ng-container *ngSwitchCase="'h1'">
                    <h1 [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </h1>
                </ng-container>
                <ng-container *ngSwitchCase="'h2'">
                    <h2 [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </h2>
                </ng-container>
                <ng-container *ngSwitchCase="'h3'">
                    <h3 [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </h3>
                </ng-container>
                <ng-container *ngSwitchCase="'h4'">
                    <h4 [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </h4>
                </ng-container>
                <ng-container *ngSwitchCase="'h5'">
                    <h5 [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </h5>
                </ng-container>
                <ng-container *ngSwitchCase="'h6'">
                    <h6 [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </h6>
                </ng-container>
                <ng-container *ngSwitchCase="'div'">
                    <div [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </div>
                </ng-container>
                <ng-container *ngSwitchCase="'span'">
                    <span [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </span>
                </ng-container>
                <ng-container *ngSwitchCase="'table'">
                    <table [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </table>
                </ng-container>
                <ng-container *ngSwitchCase="'td'">
                    <td [ngStyle]="element.style" (keydown)="onShortcut($event, element)"
                        (attr.colspan)="element.options.colspan" (attr.rowspan)="element.options.rowspan">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </td>
                </ng-container>
                <ng-container *ngSwitchCase="'th'">
                    <th [ngStyle]="element.style" (keydown)="onShortcut($event, element)"
                        (attr.colspan)="element.options.colspan" (attr.rowspan)="element.options.rowspan">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </th>
                </ng-container>
                <ng-container *ngSwitchCase="'tr'">
                    <tr [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </tr>
                </ng-container>
                <ng-container *ngSwitchCase="'tbody'">
                    <tbody [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </tbody>
                </ng-container>
                <ng-container *ngSwitchCase="'thead'">
                    <thead [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </thead>
                </ng-container>
                <ng-container *ngSwitchCase="'tfoot'">
                    <tfoot [ngStyle]="element.style" (keydown)="onShortcut($event, element)">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </tfoot>
                </ng-container>
                <ng-container *ngSwitchCase="'button'">
                    <button [ngStyle]="element.style" (keydown)="onShortcut($event, element)"
                        (click)="element.options.onClick()" [disabled]="element.options.disabled">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </button>
                </ng-container>
                <ng-container *ngSwitchCase="'drayman-input-field'">
                    <app-drayman-input-field [ngStyle]="element.style" (keydown)="onShortcut($event, element)"
                        [options]="element.options">
                        <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}">
                        </ng-container>
                    </app-drayman-input-field>
                </ng-container>
                <ng-container *ngSwitchCase="'img'">
                    <img [ngStyle]="element.style" (keydown)="onShortcut($event, element)" [src]="element.options.src"
                        [alt]="element.options.alt" />
                </ng-container>
                <!--ELEMENTS-END-->
                <ng-container *ngSwitchCase="'text'">
                    {{element.options.text}}
                </ng-container>
            </ng-container>
        </ng-container>
    </ng-container>
</ng-template>