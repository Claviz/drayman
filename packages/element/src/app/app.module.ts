import { LazyElementsModule } from '@angular-extensions/elements';
import { ClipboardModule } from '@angular/cdk/clipboard';
import { CUSTOM_ELEMENTS_SCHEMA, Injector, NgModule } from '@angular/core';
import { createCustomElement } from '@angular/elements';
import { ReactiveFormsModule } from '@angular/forms';
import { MatDialogModule } from '@angular/material/dialog';
import { MatSnackBarModule } from '@angular/material/snack-bar';
import { BrowserModule } from '@angular/platform-browser';
import { BrowserAnimationsModule, } from '@angular/platform-browser/animations';
import { ElementZoneStrategyFactory } from 'elements-zone-strategy';
// import { AppComponent } from './app.component';

import { DraymanElementComponent } from './drayman-element.component';
import { DraymanInputFieldComponent } from './input-field/drayman-input-field.component';
import { DraymanModalComponent } from './modal/drayman-modal.component';

@NgModule({
  schemas: [CUSTOM_ELEMENTS_SCHEMA],
  imports: [
    BrowserModule,
    BrowserAnimationsModule,
    LazyElementsModule,
    MatDialogModule,
    MatSnackBarModule,
    ClipboardModule,
    ReactiveFormsModule,
  ],
  declarations: [
    // AppComponent,
    DraymanElementComponent,
    DraymanModalComponent,
    DraymanInputFieldComponent,
  ],
  providers: [],
  bootstrap: [],
  entryComponents: [DraymanElementComponent],
})
export class AppModule {
  constructor(private injector: Injector) {
  }

  ngDoBootstrap() {
    const strategyFactory = new ElementZoneStrategyFactory(DraymanElementComponent, this.injector);
    const elm = createCustomElement(DraymanElementComponent, { injector: this.injector, strategyFactory });
    customElements.define('drayman-element', elm);
  }

}