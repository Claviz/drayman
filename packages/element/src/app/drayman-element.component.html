<ng-container *ngIf="viewTree">
    <!-- <ng-container *ngFor="let element of viewTree; trackBy: trackByFn"> -->
    <ng-container *ngTemplateOutlet="childrenTemplate; context: {children: viewTree}"></ng-container>
    <!-- </ng-container> -->
</ng-container>
<ng-template #childrenTemplate let-children="children">
    <ng-container *ngFor="let element of children; trackBy: trackByFn">
        <ng-container *ngIf="!isHtmlElement(element.type)">
            <ax-lazy-element [ngStyle]="element.style"
                *axLazyElementDynamic="element.type; url: config?.elementUrl + element.type;"
                [options]="element.options">
            </ax-lazy-element>
        </ng-container>
        <ng-container *ngIf="isHtmlElement(element.type)">
            <ng-container [ngSwitch]="element.type">
                <!--ELEMENTS-START-->
<ng-container *ngSwitchCase="'i'">
<i [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</i>
</ng-container>
<ng-container *ngSwitchCase="'b'">
<b [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</b>
</ng-container>
<ng-container *ngSwitchCase="'p'">
<p [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</p>
</ng-container>
<ng-container *ngSwitchCase="'h1'">
<h1 [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</h1>
</ng-container>
<ng-container *ngSwitchCase="'h2'">
<h2 [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</h2>
</ng-container>
<ng-container *ngSwitchCase="'h3'">
<h3 [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</h3>
</ng-container>
<ng-container *ngSwitchCase="'h4'">
<h4 [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</h4>
</ng-container>
<ng-container *ngSwitchCase="'h5'">
<h5 [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</h5>
</ng-container>
<ng-container *ngSwitchCase="'h6'">
<h6 [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</h6>
</ng-container>
<ng-container *ngSwitchCase="'td'">
<td [attr.colspan]="element.options.colspan" [attr.rowspan]="element.options.rowspan" [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</td>
</ng-container>
<ng-container *ngSwitchCase="'div'">
<div [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</div>
</ng-container>
<ng-container *ngSwitchCase="'span'">
<span [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</span>
</ng-container>
<ng-container *ngSwitchCase="'table'">
<table [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</table>
</ng-container>
<ng-container *ngSwitchCase="'tbody'">
<tbody [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</tbody>
</ng-container>
<ng-container *ngSwitchCase="'thead'">
<thead [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</thead>
</ng-container>
<ng-container *ngSwitchCase="'tfoot'">
<tfoot [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</tfoot>
</ng-container>
<ng-container *ngSwitchCase="'tr'">
<tr [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</tr>
</ng-container>
<ng-container *ngSwitchCase="'th'">
<th [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</th>
</ng-container>
<ng-container *ngSwitchCase="'input'">
<app-drayman-input-field [options]="element.options" [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</app-drayman-input-field>
</ng-container>
<ng-container *ngSwitchCase="'img'">
<img [src]="domSanitizer.bypassSecurityTrustUrl(element.options.src)" [alt]="element.options.alt" [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)" />
</ng-container>
<ng-container *ngSwitchCase="'button'">
<button (click)="element.options.onClick()" [disabled]="element.options.disabled" [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</button>
</ng-container>
<ng-container *ngSwitchCase="'pre'">
<pre [class]="element.options.class" [ngStyle]="element.style" [attr.tabindex]="element.options.tabindex" (keydown)="onShortcut($event, element)">
<ng-container *ngTemplateOutlet="childrenTemplate; context: {children: element.children}"></ng-container>
</pre>
</ng-container>
                <!--ELEMENTS-END-->
                <ng-container *ngSwitchCase="'text'">
                    {{element.options.text}}
                </ng-container>
            </ng-container>
        </ng-container>
    </ng-container>
</ng-template>