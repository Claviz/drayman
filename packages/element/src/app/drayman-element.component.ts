import { Clipboard } from '@angular/cdk/clipboard';
import {
  ChangeDetectorRef,
  Component,
  Input,
  NgZone,
  OnChanges,
  OnDestroy,
  OnInit,
  Optional,
  SimpleChanges,
} from '@angular/core';
import { MatDialog, MatDialogRef } from '@angular/material/dialog';
import { MatSnackBar } from '@angular/material/snack-bar';
import { applyPatch } from 'fast-json-patch';
import isHotkey from 'is-hotkey';

import { IConnectionService } from './interfaces/connection-service';
import { DraymanModalComponent } from './modal/drayman-modal.component';

@Component({
  selector: 'app-drayman-element',
  templateUrl: './drayman-element.component.html',
  styleUrls: ['./drayman-element.component.scss'],
})
export class DraymanElementComponent implements OnChanges, OnInit, OnDestroy {
  @Input() component?: string;
  @Input() options: any;
  @Input() config?: { connection: IConnectionService, elementUrl: string };
  @Input() isModal = false;

  componentInstanceId: any;
  view = {};
  previouslySerializedTree = [];
  viewTree: any[] = [];

  constructor(
    @Optional() private dialogRef: MatDialogRef<DraymanModalComponent>,
    private dialog: MatDialog,
    private snackBar: MatSnackBar,
    private clipboard: Clipboard,
    private ref: ChangeDetectorRef,
    private ngZone: NgZone,
  ) { }

  ngOnChanges(changes: SimpleChanges) {
    if (changes.options && !changes.options.firstChange) {
      this.config?.connection.updateComponentInstanceProps({ componentInstanceId: this.componentInstanceId, options: this.options });
    }
  }

  trackByFn(index: number, item: any) {
    return item.key;
  }

  ngOnDestroy() {
    this.config?.connection.destroyComponentInstance({ componentInstanceId: this.componentInstanceId });
  }

  traverseTree(tree: any) {
    for (let item of tree) {
      const elementOptions = item.options;
      if (elementOptions) {
        for (const eventName of Object.keys(elementOptions)) {
          if (isEvent(eventName)) {
            elementOptions[eventName] = this.emit.bind(this, `${item.key}/${eventName}`);
          }
        }
      }
      this.traverseTree(item.children);
    }
  }

  async ngOnInit() {
    while (!this.config) {
      await new Promise(resolve => setTimeout(resolve, 100));
    }
    this.componentInstanceId = await this.config.connection.initializeComponent({ componentId: this.component, componentOptions: this.options, location: { href: window.location.href }, isModal: this.isModal });
    this.config.connection.onEvent(this.componentInstanceId, ({ type, payload }) => {
      this.ngZone.run(() => {
        if (type === 'closeModal') {
          if (this.dialogRef) {
            const { data } = payload;
            this.dialogRef.close(data);
          }
        } else if (type === 'view') {
          applyPatch(this.previouslySerializedTree, (payload.view || []));
          const tree = JSON.parse(JSON.stringify(this.previouslySerializedTree));
          this.traverseTree(tree);
          this.viewTree = tree;
          this.ref.detectChanges();
        }
        else if (type === 'openModal') {
          const { component, options, onCloseCallbackId, modalOptions } = payload;
          const width = modalOptions?.width || '50vw';
          const height = modalOptions?.height || '80vh';
          const dialog = this.dialog.open(DraymanModalComponent, {
            data: {
              component,
              options,
              config: this.config,
            },
            height,
            width,
            maxHeight: height,
            maxWidth: width,
          });
          dialog.afterClosed().subscribe((data) => {
            if (onCloseCallbackId) {
              this.config?.connection.handleBrowserCallback({ callbackId: onCloseCallbackId, data });
            }
          });
        } else if (type === 'openWindow') {
          const { windowName, url, windowFeatures } = payload;
          window.open(url, windowName, windowFeatures);
        } else if (type === 'copyToClipboard') {
          const { text } = payload;
          this.clipboard.copy(text);
          this.snackBar.open(`Copied to clipboard!`, undefined, { duration: 2500 });
        } else if (type === 'openSnackBar') {
          let { options, message: snackBarMessage, onCloseCallbackId } = payload;
          let { duration, horizontalPosition, verticalPosition, action } = options || {};
          const snackbar = this.snackBar.open(snackBarMessage, action, { duration, horizontalPosition, verticalPosition });
          if (onCloseCallbackId) {
            snackbar.afterDismissed().subscribe((data) => {
              this.config?.connection.handleBrowserCallback({ callbackId: onCloseCallbackId, data: { dismissedByAction: data.dismissedByAction } });
            })
          }
        } else if (type === 'navigate') {
          const { path } = payload;
          this.config?.connection.navigate(path);
        }
      });
    });
  }

  emit = async (eventName: string, options = {}, files: any[] = []) => {
    const formData = new FormData();
    formData.append('eventName', eventName);
    formData.append('componentInstanceId', this.componentInstanceId);
    formData.append('options', JSON.stringify(options));
    for (let file of files) {
      formData.append(file.fieldName, file.file, file.fileName);
    }
    return await this.config?.connection.postFormData(formData);
  }

  onShortcut($event: KeyboardEvent, element: any) {
    for (const attr of Object.keys(element.options)) {
      if (attr.startsWith('onShortcut')) {
        const shortcutSplit = attr.split(':');
        if (shortcutSplit.length === 2) {
          const shortcut = shortcutSplit[1];
          if (isHotkey(shortcut, $event)) {
            $event.preventDefault();
            element.options[attr]();
          }
        }
      }
    }
  }

  isHtmlElement(name: string) {
    return name === name.toLowerCase() && !name.includes('-');
  }
}

const isEvent = (optionName: string) => optionName?.length > 2 && optionName.slice(0, 2) === 'on' && optionName[2] === optionName[2].toUpperCase();
