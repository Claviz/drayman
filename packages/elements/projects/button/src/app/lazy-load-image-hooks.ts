import { Attributes, IntersectionObserverHooks, LAZYLOAD_IMAGE_HOOKS, LazyLoadImageModule } from 'ng-lazyload-image';

export class LazyLoadImageHooks extends IntersectionObserverHooks {
    async loadImage({ imagePath }: Attributes): Promise<string> {
        const response = await fetch(imagePath);
        const blob = await response.blob();
        if (blob.size) {
            const url = URL.createObjectURL(blob);
            return url;
        }
    }
}