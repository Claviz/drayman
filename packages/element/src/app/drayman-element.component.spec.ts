jest.mock('fast-json-patch');
import { Clipboard } from '@angular/cdk/clipboard';
import { MatDialog, MatDialogRef } from '@angular/material/dialog';
import { MatSnackBar } from '@angular/material/snack-bar';
import { Spectator, createComponentFactory } from '@ngneat/spectator/jest';
import { AppModule } from './app.module';
import { DraymanElementComponent } from './drayman-element.component';
import { DraymanModalComponent } from './modal/drayman-modal.component';
import { applyPatch } from 'fast-json-patch';
import { Subject } from 'rxjs';

describe('DraymanElementComponent', () => {

  let spectator: Spectator<DraymanElementComponent>;
  const createComponent = createComponentFactory({
    component: DraymanElementComponent,
    imports: [AppModule],
    declareComponent: false,
    mocks: [Clipboard, MatSnackBar, MatDialogRef, MatDialog]
  });
  let eventHandler: ((event: { type: string, payload: any }) => void) | undefined;
  (applyPatch as jest.Mock).mockImplementation((previousTree, currentTree) => {
    previousTree[0] = currentTree[0];
  });
  const mockInitializeComponent = jest.fn(async (options) => {
    return '123';
  });
  const mockDestroyComponentInstance = jest.fn((componentInstanceId) => {
  });
  const mockHandleBrowserCallback = jest.fn((options) => {
  });
  const mockNavigate = jest.fn((path) => {
  });
  const mockOnEvent = jest.fn((componentInstanceId, handler) => {
    eventHandler = handler;
  });
  const mockPostFormData = jest.fn(async (formData) => {
  });
  const mockUpdateComponentInstanceProps = jest.fn((options) => {
  });
  const config = {
    elementUrl: '',
    connection: {
      initializeComponent: mockInitializeComponent,
      destroyComponentInstance: mockDestroyComponentInstance,
      handleBrowserCallback: mockHandleBrowserCallback,
      navigate: mockNavigate,
      onEvent: mockOnEvent,
      postFormData: mockPostFormData,
      updateComponentInstanceProps: mockUpdateComponentInstanceProps,
    }
  }

  beforeEach(() => {
    eventHandler = undefined;
    jest.clearAllMocks();
  });

  test(`should call connection mockInitializeComponent on init`, async () => {
    spectator = createComponent({
      props: {
        component: 'hello-world',
        options: { text: 'hi!' },
        config,
      }
    });
    await spectator.fixture.whenStable();

    expect(mockInitializeComponent.mock.calls.length).toEqual(1);
    expect(mockInitializeComponent.mock.calls[0][0]).toEqual({
      componentId: 'hello-world',
      componentOptions: { text: 'hi!' },
      location: { href: 'http://localhost/' }
    });
  });

  test(`should call connection updateComponentInstanceProps on options changes`, async () => {
    spectator = createComponent({
      props: {
        component: 'hello-world',
        options: { text: 'hi!' },
        config,
      }
    });
    await spectator.fixture.whenStable();

    expect(mockUpdateComponentInstanceProps.mock.calls.length).toEqual(0);
    spectator.setInput({
      options: { text: 'bye!' }
    })
    expect(mockUpdateComponentInstanceProps.mock.calls.length).toEqual(1);
    expect(mockUpdateComponentInstanceProps.mock.calls[0]).toEqual([
      { componentInstanceId: '123', options: { text: 'bye!' } }
    ]);
  });

  test(`should retry while config not received`, async () => {
    spectator = createComponent({
      props: {
        component: 'hello-world',
        options: { text: 'hi!' },
      }
    });

    expect(mockInitializeComponent.mock.calls.length).toEqual(0);
    spectator.setInput({
      config,
    })

    await spectator.fixture.whenStable();

    expect(mockInitializeComponent.mock.calls.length).toEqual(1);
  });

  test(`should register event handler with the received component id`, async () => {
    spectator = createComponent({
      props: {
        component: 'hello-world',
        options: { text: 'hi!' },
        config,
      }
    });
    await spectator.fixture.whenStable();
    expect(mockOnEvent.mock.calls.length).toEqual(1);
    expect(mockOnEvent.mock.calls[0][0]).toEqual('123');
    expect(mockOnEvent.mock.calls[0][1]).toEqual(eventHandler);
  });

  describe(`events`, () => {
    test(`should copy to clipboard`, async () => {
      spectator = createComponent({
        props: {
          component: 'hello-world',
          options: { text: 'hi!' },
          config,
        }
      });
      await spectator.fixture.whenStable();
      eventHandler?.({
        type: 'copyToClipboard',
        payload: {
          text: 'hello, world!'
        }
      });
      const clipboardService = spectator.inject<Clipboard>(Clipboard);
      expect(clipboardService.copy.mock.calls.length).toEqual(1);
      expect(clipboardService.copy.mock.calls[0][0]).toEqual('hello, world!');
      const snackBarService = spectator.inject<MatSnackBar>(MatSnackBar);
      expect(snackBarService.open.mock.calls.length).toEqual(1);
      expect(snackBarService.open.mock.calls[0]).toEqual([
        `Copied to clipboard!`, undefined, { duration: 2500 }
      ]);
    });

    test(`should navigate`, async () => {
      spectator = createComponent({
        props: {
          component: 'hello-world',
          options: { text: 'hi!' },
          config,
        }
      });
      await spectator.fixture.whenStable();
      eventHandler?.({
        type: 'navigate',
        payload: {
          path: '/home'
        }
      });
      expect(mockNavigate.mock.calls.length).toEqual(1);
      expect(mockNavigate.mock.calls[0][0]).toEqual('/home');
    });

    test(`should open window`, async () => {
      spectator = createComponent({
        props: {
          component: 'hello-world',
          options: { text: 'hi!' },
          config,
        }
      });
      await spectator.fixture.whenStable();
      const openWindowMock = jest.fn();
      window.open = openWindowMock;
      eventHandler?.({
        type: 'openWindow',
        payload: { windowName: '', url: 'google.com', windowFeatures: {} }
      });
      expect(openWindowMock.mock.calls.length).toEqual(1);
      expect(openWindowMock.mock.calls[0]).toEqual(['google.com', '', {}]);
    });

    test(`should close modal if current component is opened in modal`, async () => {
      spectator = createComponent({
        props: {
          component: 'hello-world',
          options: { text: 'hi!' },
          config,
        }
      });
      await spectator.fixture.whenStable();
      eventHandler?.({
        type: 'closeModal',
        payload: { data: 'bye!' }
      });
      const dialogRef = spectator.inject<MatDialogRef<DraymanModalComponent>>(MatDialogRef);
      expect(dialogRef.close.mock.calls.length).toEqual(1);
      expect(dialogRef.close.mock.calls[0]).toEqual(['bye!']);
    });

    test(`should open and close modal`, async () => {
      spectator = createComponent({
        props: {
          component: 'hello-world',
          options: { text: 'hi!' },
          config,
        }
      });
      await spectator.fixture.whenStable();
      const payload = { component: 'dialog-component', options: {}, onCloseCallbackId: 'close-dialog-component' };
      const matDialog = spectator.inject<MatDialog>(MatDialog);
      const closeModal = new Subject();
      matDialog.open.mockImplementation(() => {
        return {
          afterClosed: () => closeModal
        } as any
      });
      eventHandler?.({
        type: 'openModal',
        payload,
      });
      expect(matDialog.open.mock.calls.length).toEqual(1);
      expect(matDialog.open.mock.calls[0]).toEqual([
        DraymanModalComponent, {
          data: {
            component: payload.component,
            options: payload.options,
            config,
          }
        }
      ]);
      closeModal.next('bye!');
      expect(mockHandleBrowserCallback.mock.calls.length).toEqual(1);
      expect(mockHandleBrowserCallback.mock.calls[0]).toEqual([{
        callbackId: 'close-dialog-component',
        data: 'bye!'
      }]);
    });

    test(`should open and close snack bar`, async () => {
      spectator = createComponent({
        props: {
          component: 'hello-world',
          options: { text: 'hi!' },
          config,
        }
      });
      await spectator.fixture.whenStable();
      const payload = {
        message: 'surprise!',
        options: {
          duration: 2000, horizontalPosition: 'center', verticalPosition: 'bottom', action: 'close'
        },
        onCloseCallbackId: 'close-surprise',
      };
      const matSnackBar = spectator.inject<MatSnackBar>(MatSnackBar);
      const closeSnackBar = new Subject();
      matSnackBar.open.mockImplementation(() => {
        return {
          afterDismissed: () => closeSnackBar,
        } as any
      });
      eventHandler?.({
        type: 'openSnackBar',
        payload,
      });
      expect(matSnackBar.open.mock.calls.length).toEqual(1);
      expect(matSnackBar.open.mock.calls[0]).toEqual(['surprise!', 'close', { duration: 2000, horizontalPosition: 'center', verticalPosition: 'bottom' }]);
      closeSnackBar.next({ dismissedByAction: true });
      expect(mockHandleBrowserCallback.mock.calls.length).toEqual(1);
      expect(mockHandleBrowserCallback.mock.calls[0]).toEqual([{
        callbackId: 'close-surprise',
        data: { dismissedByAction: true }
      }]);
    });

    test(`should apply view tree`, async () => {
      spectator = createComponent({
        props: {
          component: 'hello-world',
          options: { text: 'hi!' },
          config,
        }
      });
      await spectator.fixture.whenStable();
      const emitMock = jest.fn();
      spectator.component.emit.bind = emitMock;
      const tree = [{
        type: 'button',
        options: {
          onClick: true,
        },
        key: 'btn-1',
        children: [],
      }];
      eventHandler?.({
        type: 'view',
        payload: { view: tree }
      });
      expect([{ ...spectator.component.viewTree[0], options: {} }]).toEqual([{ ...tree[0], options: {} }]);
      expect(emitMock.mock.calls[0][1]).toEqual('btn-1/onClick');
    });
  });

  test(`should render button`, async () => {
    spectator = createComponent({
      props: {
        component: 'hello-world',
        options: { text: 'hi!' },
        config,
      }
    });
    await spectator.fixture.whenStable();
    const tree = [{
      type: 'button',
      options: {
        onClick: true,
      },
      key: 'btn-1',
      children: [],
    }];
    const emitMock = jest.fn();
    spectator.component.emit = emitMock;
    eventHandler?.({
      type: 'view',
      payload: { view: tree }
    });

    expect(spectator.fixture).toMatchSnapshot();

    spectator.click('button');
    await spectator.fixture.whenStable();
    expect(emitMock.mock.calls[0]).toEqual(['btn-1/onClick']);
  });

});
