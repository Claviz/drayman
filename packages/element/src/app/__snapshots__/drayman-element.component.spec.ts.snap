// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraymanElementComponent should render button 1`] = `
<app-drayman-element
  clipboard={[Function Object]}
  component={[Function String]}
  componentInstanceId={[Function String]}
  config={[Function Object]}
  dialog={[Function Object]}
  dialogRef={[Function Object]}
  emit={[Function Function]}
  isModal="false"
  ngZone={[Function NgZone]}
  options={[Function Object]}
  previouslySerializedTree={[Function Array]}
  ref={[Function ViewRef_]}
  snackBar={[Function Object]}
  view={[Function Object]}
  viewTree={[Function Array]}
>
  <button>
    
    
    
  </button>
</app-drayman-element>
`;
